### ISSUE TITLE
[MPS] F.linear with bias silently returns wrong results when 3D batch dim exceeds 2^16 (65536)

### ISSUE THREAD

**Issue by bogatyy** [pytorch/pytorch#189495]
### 🐛 Describe the bug

`F.linear` (and `nn.Linear`) with a **bias** on a **3D input** silently returns wrong results on MPS when the leading batch dimension exceeds **65536 (2^16)**. The corruption is silent (no error/warning), affects a subset of batch rows, and appears **nondeterministic** across identical calls.

```python
import torch
import torch.nn.functional as F

torch.manual_seed(0)
x = torch.randn(65537, 11, 128)
w = torch.randn(512, 128)
b = torch.randn(512)

ref = F.linear(x, w, b)
out = F.linear(x.to("mps"), w.to("mps"), b.to("mps")).cpu()

print((out - ref).abs().max())  # ~60 (!) — expected ~1e-5
print((~torch.isclose(out, ref, atol=1e-3, rtol=1e-3)).any(-1).any(-1).sum().item(), "corrupted rows")
```

Reproduces identically on **2.12.0, 2.13.0 (latest stable), and nightly 2.14.0.dev20260709**.

Each of the following makes the result exactly correct:

- `B <= 65536` (the threshold is exactly 65536 → 65537)
- `bias=None` (e.g. `nn.Linear(..., bias=False)`)
- reshaping the input to 2D first: `F.linear(x.reshape(-1, 128), w, b)`
- fp16/bf16 input (on M5), or a non-contiguous input — see root cause for why

### Root cause

The corruption is a **16-bit batch-index wraparound inside Apple's `MPSNDArrayMatrixMultiplication` fused 3-source (A@B+C) kernel**: batch `b` is written to output slot `b mod 65536`. PyTorch's `_mps_linear_nograph` (aten/src/ATen/native/mps/operations/Linear.mm) hits it because the bias path encodes that kernel with `sourceCount:3` on the unflattened 3D NDArray.

Evidence, all on Apple M5 Max / macOS 26.5.2:

1. **Output forensics (through PyTorch).** At `B=65537`, output row 0 equals row 65536's *correct* result to float precision (max diff 9.5e-6), i.e. batch 65536 wrapped onto slot 0.
2. **Pure Objective-C repro, no PyTorch** (attached below), mirroring Linear.mm's NDArray setup:
   ```
   B=  65536 bias=1  bad rows:      0
   B=  65537 bias=1  bad rows:     65  first bad: 0  row0==row65536's result: YES
   B=  80000 bias=1  bad rows:  28928  first bad: 0  row0==row65536's result: YES
   B=  80000 bias=0  bad rows:      0
   ```
   The 28,928 count is exactly `2 × (80000 − 65536)`: each wrapped batch clobbers one victim row below 65536 **and leaves its own output row unwritten**. (The ObjC repro poisons the output buffer to show the unwritten rows; through PyTorch those rows expose recycled heap memory, which is why results look nondeterministic and allocator-state-dependent.) The corrupted-row count through PyTorch at B=80000 is also exactly 28,928.
3. **The 2-source (no-bias) variant of the same kernel is correct** at all sizes tested — hence `bias=False` being clean.
4. **Path analysis** explains every workaround: fp16/bf16 on M5 is clean because `needs_nd_workaround()` (added for #180776 — a *different* known defect in this same kernel family on M5) flattens those inputs to 2D before encoding; non-contiguous inputs are clean because they fall back to the MPSGraph path; 2D inputs are clean because the wraparound is in the NDArray batch dimension, not matrix rows (a 2D input with 721,907 rows is exact).

Related precedent: the MPS convolution kernel had a 2^16 output-channel limit, guarded with `TORCH_CHECK_NOT_IMPLEMENTED` and fixed by Apple in macOS 15.1 (#129207, #142515). The now-closed #122045 (macOS-13-only `F.linear` corruption) is a different bug on a code path that no longer exists.

### Suggested fix

For a contiguous input with `bias.dim() <= 1`, `input.flatten(0, -2)` is a free view and sidesteps the broken kernel entirely — the machinery already exists in Linear.mm as the #180776 workaround. Extending that flatten from "fp16/bf16 on Apple10+" to also cover "leading-dims product > 65536" (any dtype) fixes this with no perf cost. Alternatively guard like Convolution.mm does, but the silent fix seems strictly better here. Note `PYTORCH_MPS_PREFER_METAL` does not reroute `F.linear` (#188728), so users currently have no opt-out knob. This should presumably also be reported to Apple via Feedback Assistant, since the defect is in the MPS framework kernel itself.

### How this was found

Evaluating a small transformer on an 80k-example dataset in a single batch: train-set metrics looked mysteriously *worse* than test-set metrics (40k rows) because only the train eval crossed the 2^16 threshold — the model's FeedForward `nn.Linear` layers (bias=True) corrupted ~36% of rows.

### Minimal pure-MPS reproducer (no PyTorch)

<details>
<summary><code>mps_ndarray_repro.m</code> — creates fp32 MPSNDArrays input (B,T,K), weight (N,K) with <code>transposeDimension:0:1</code> + <code>preferPackedRows</code> (exactly as Linear.mm does), bias (N); encodes <code>MPSNDArrayMatrixMultiplication</code> with <code>sourceCount:3</code>; validates against CPU arithmetic. B=65537 corrupts, B=65536 exact; <code>sourceCount:2</code> exact at any B.</summary>

```objc
// Pure-MPS repro: MPSNDArrayMatrixMultiplication (3-source, A@B+C) writes batch
// index b to b mod 65536 when the leading dim of a 3D source exceeds 2^16.
// Mirrors PyTorch's _mps_linear_nograph setup (Linear.mm).
//
// Build: clang -fobjc-arc -framework Foundation -framework Metal \
//        -framework MetalPerformanceShaders mps_ndarray_repro.m -o mps_ndarray_repro
#import <Foundation/Foundation.h>
#import <Metal/Metal.h>
#import <MetalPerformanceShaders/MetalPerformanceShaders.h>

static const NSUInteger T = 2, K = 32, N = 32;

static float inputVal(NSUInteger b, NSUInteger t, NSUInteger k) {
  return (float)((b * 31 + t * 7 + k * 3) % 13) - 6.0f;
}
static float weightVal(NSUInteger n, NSUInteger k) {  // stored (N, K), used as K x N
  return (float)((n * 5 + k) % 7) - 3.0f;
}
static float biasVal(NSUInteger n) { return (float)(n % 5) - 2.0f; }

static float expectedOut(NSUInteger b, NSUInteger t, NSUInteger n) {
  float acc = biasVal(n);
  for (NSUInteger k = 0; k < K; k++) acc += inputVal(b, t, k) * weightVal(n, k);
  return acc;
}

static void runCase(id<MTLDevice> dev, id<MTLCommandQueue> queue, NSUInteger B, BOOL withBias) {
  id<MTLBuffer> inBuf = [dev newBufferWithLength:B * T * K * 4 options:MTLResourceStorageModeShared];
  id<MTLBuffer> wBuf = [dev newBufferWithLength:N * K * 4 options:MTLResourceStorageModeShared];
  id<MTLBuffer> bBuf = [dev newBufferWithLength:N * 4 options:MTLResourceStorageModeShared];
  id<MTLBuffer> outBuf = [dev newBufferWithLength:B * T * N * 4 options:MTLResourceStorageModeShared];

  float *in = inBuf.contents, *w = wBuf.contents, *bias = bBuf.contents, *out = outBuf.contents;
  for (NSUInteger b = 0; b < B; b++)
    for (NSUInteger t = 0; t < T; t++)
      for (NSUInteger k = 0; k < K; k++) in[(b * T + t) * K + k] = inputVal(b, t, k);
  for (NSUInteger n = 0; n < N; n++)
    for (NSUInteger k = 0; k < K; k++) w[n * K + k] = weightVal(n, k);
  for (NSUInteger n = 0; n < N; n++) bias[n] = biasVal(n);
  memset(out, 0xAA, B * T * N * 4);  // poison so untouched rows are detectable

  // shapes follow MPSNDArrayDescriptor convention (fastest-moving dim first),
  // matching PyTorch's getMPSShape(): input (B, T, K) -> @[K, T, B] reversed? No:
  // MPS wants sizes in the same order as PyTorch passes them via getMPSShape,
  // which lists them slowest-first. descriptorWithDataType:shape: takes an
  // NSArray in row-major (slowest-first) order.
  MPSNDArrayDescriptor *inDesc =
      [MPSNDArrayDescriptor descriptorWithDataType:MPSDataTypeFloat32
                                             shape:@[ @(B), @(T), @(K) ]];
  MPSNDArray *inArr = [[MPSNDArray alloc] initWithBuffer:inBuf offset:0 descriptor:inDesc];

  MPSNDArrayDescriptor *wDesc =
      [MPSNDArrayDescriptor descriptorWithDataType:MPSDataTypeFloat32
                                             shape:@[ @(N), @(K) ]];
  wDesc.preferPackedRows = YES;
  [wDesc transposeDimension:0 withDimension:1];  // logical (K, N), like PyTorch
  MPSNDArray *wArr = [[MPSNDArray alloc] initWithBuffer:wBuf offset:0 descriptor:wDesc];

  MPSNDArrayDescriptor *bDesc =
      [MPSNDArrayDescriptor descriptorWithDataType:MPSDataTypeFloat32 shape:@[ @(N) ]];
  MPSNDArray *bArr = [[MPSNDArray alloc] initWithBuffer:bBuf offset:0 descriptor:bDesc];

  MPSNDArrayDescriptor *outDesc =
      [MPSNDArrayDescriptor descriptorWithDataType:MPSDataTypeFloat32
                                             shape:@[ @(B), @(T), @(N) ]];
  MPSNDArray *outArr = [[MPSNDArray alloc] initWithBuffer:outBuf offset:0 descriptor:outDesc];

  MPSNDArrayMatrixMultiplication *kernel =
      [[MPSNDArrayMatrixMultiplication alloc] initWithDevice:dev sourceCount:withBias ? 3 : 2];

  id<MTLCommandBuffer> cmdBuf = [queue commandBuffer];
  id<MTLComputeCommandEncoder> enc = [cmdBuf computeCommandEncoder];
  NSArray *sources = withBias ? @[ inArr, wArr, bArr ] : @[ inArr, wArr ];
  [kernel encodeToCommandEncoder:enc commandBuffer:cmdBuf sourceArrays:sources destinationArray:outArr];
  [enc endEncoding];
  [cmdBuf commit];
  [cmdBuf waitUntilCompleted];

  NSUInteger badRows = 0, firstBad = NSUIntegerMax;
  BOOL row0HasRow65536 = NO;
  for (NSUInteger b = 0; b < B; b++) {
    BOOL bad = NO;
    for (NSUInteger t = 0; t < T && !bad; t++)
      for (NSUInteger n = 0; n < N && !bad; n++) {
        float exp = expectedOut(b, t, n) - (withBias ? 0.0f : biasVal(n));
        if (fabsf(out[(b * T + t) * N + n] - exp) > 1e-2f) bad = YES;
      }
    if (bad) {
      badRows++;
      if (firstBad == NSUIntegerMax) firstBad = b;
    }
  }
  if (B > 65536) {
    row0HasRow65536 = YES;
    for (NSUInteger t = 0; t < T && row0HasRow65536; t++)
      for (NSUInteger n = 0; n < N && row0HasRow65536; n++) {
        float exp = expectedOut(65536, t, n) - (withBias ? 0.0f : biasVal(n));
        if (fabsf(out[t * N + n] - exp) > 1e-2f) row0HasRow65536 = NO;
      }
  }
  printf("B=%7lu bias=%d  bad rows: %6lu  first bad: %ld  row0==row65536's result: %s\n",
         (unsigned long)B, withBias, (unsigned long)badRows,
         firstBad == NSUIntegerMax ? -1L : (long)firstBad,
         B > 65536 ? (row0HasRow65536 ? "YES" : "no") : "n/a");
}

int main(void) {
  @autoreleasepool {
    id<MTLDevice> dev = MTLCreateSystemDefaultDevice();
    id<MTLCommandQueue> queue = [dev newCommandQueue];
    printf("device: %s\n", dev.name.UTF8String);
    runCase(dev, queue, 65536, YES);
    runCase(dev, queue, 65537, YES);
    runCase(dev, queue, 80000, YES);
    runCase(dev, queue, 80000, NO);
  }
  return 0;
}
```

</details>

Build & run:
```
clang -fobjc-arc -framework Foundation -framework Metal \
      -framework MetalPerformanceShaders mps_ndarray_repro.m -o repro && ./repro
```

Output on Apple M5 Max, macOS 26.5.2:
```
B=  65536 bias=1  bad rows:      0  first bad: -1  row0==row65536's result: n/a
B=  65537 bias=1  bad rows:     65  first bad: 0  row0==row65536's result: YES
B=  80000 bias=1  bad rows:  28928  first bad: 0  row0==row65536's result: YES
B=  80000 bias=0  bad rows:      0  first bad: -1  row0==row65536's result: no
```

### Versions

- PyTorch: 2.12.0, 2.13.0, 2.14.0.dev20260709 (all reproduce; arm64 wheels)
- macOS: 26.5.2 (build 25F84)
- Hardware: Apple M5 Max
- Python: 3.12.13

cc @ezyang @gchanan @kadeng @msaroufim @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @malfet @aditvenk @jianyuh @nikitaved @Lezcano @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01 @robert-hardwick @nWEIdia @kulinseth @DenisVieriu97 @jhavukainen @Isalia20

**Comment by bogatyy** [comment]
Reported the underlying kernel defect to Apple as FB23663591 (Feedback Assistant), with the standalone `MPSNDArrayMatrixMultiplication` reproducer from this issue attached.

A PR with a PyTorch-side workaround (extending the existing flatten-to-2D path in `Linear.mm`) is prepared and will be posted shortly; if a future macOS fixes the kernel, the guard can be version-scoped the way the Conv 2^16 guard was after macOS 15.1.

**Comment by eladc4** [comment]
I can reproduce a likely related manifestation on Apple Silicon MPS where two consecutive `nn.Linear` eval forwards on the exact same input and weights produce different outputs.

This is not CPU-vs-MPS numerical drift: it is MPS disagreeing with itself within the same process.

Minimal repro:
```python
# repro creates CPU input/weights, copies both to MPS, then runs:
# y1 = linear(x)
# y2 = linear(x)
# diff = (y1.cpu() - y2.cpu()).abs()
```

Reproduces on macOS 26.5.2 arm64, Apple Silicon, Python 3.12.10, torch=2.13.0 (and also nightly)

**Comment by bogatyy** [comment]
@eladc4 Thanks! This is very likely the same bug, seen from a different angle. **Could you post the actual repro (exact shapes, dtype, and code)? The key question is whether your input has `size(-3) > 65536`.**

Why self-disagreement is the expected signature of this bug: the kernel never writes the victim rows (`b >= 65536`) at all — batch `b` lands on slot `b mod 2^16` instead — so those output rows expose whatever the MPS heap last held at that address. Two consecutive calls agree or disagree purely based on allocator state. In fact, a naive back-to-back `y1 = linear(x); y2 = linear(x)` comes out identical on my machine (both calls recycle equally-stale buffers); the disagreement appears once other allocations dirty the heap between the calls, which any real workload does. Demonstration:

```python
import torch, torch.nn as nn
lin = nn.Linear(32, 64).to("mps").eval()
x = torch.randn(65537, 17, 32, device="mps")

def churn(val):  # leave recognizable garbage in the MPS heap
    for i in range(8):
        t = torch.full((65537, 17, 64), val + i, device="mps"); t.sum().item(); del t

with torch.no_grad():
    churn(100.0); y1 = lin(x).cpu()
    churn(900.0); y2 = lin(x).cpu()
print((y1 - y2).abs().max())          # 800.0
print(y1[65536, 0, :4], y2[65536, 0, :4])   # [107., 107., ...] vs [907., 907., ...]
```

Row 65536 of `y1` reads back `107.0` and of `y2` reads `907.0`, which are the fill values of the previously freed tensors, not a computation at all.

If your case is this bug, any of the following should make your `y1 == y2` exactly: `bias=False`, batch ≤ 65536, flattening the input to 2D, or a build with #189496. **If instead your `size(-3)` is ≤ 65536, please do share the full script.**
